packages/microsoft_kiota_azure/lib/src/device_code_credential.dart

```diff
@@ -0,0 +1,157 @@
+  static const String defaultTenantId = 'common';
+  final String clientId;
+  final String tenantId;
+  final String host;
+  final http.Client _httpClient;
+  final void Function(DeviceCodeInfo) challengeConsumer;
+  final Map<String, AccessToken> _tokenCache = HashMap<String, AccessToken>();
+  @override
+  Future<AccessToken> getToken(TokenRequestContext requestContext) async {
+    final cachedToken = _getCachedToken(requestContext);
+    if (cachedToken != null) {
+      return cachedToken;
+    }
+    final deviceCodeInfo = await _getDeviceCodeInfo(requestContext);
+    challengeConsumer(deviceCodeInfo);
+    final accessToken = await _pollForToken(deviceCodeInfo);
+    addTokenToCache(requestContext, accessToken);
+    return accessToken;
+  }
+
+  void addTokenToCache(TokenRequestContext requestContext, AccessToken token) {
+    _tokenCache[_getCacheKey(requestContext)] = token;
+  }
+
+  AccessToken? _getCachedToken(TokenRequestContext requestContext) {
+    final cacheKey = _getCacheKey(requestContext);
+    final cachedToken = _tokenCache[cacheKey];
+    if (cachedToken != null) {
+      if (cachedToken.expiresOn?.isAfter(DateTime.now()) ?? false) {
+        return cachedToken;
+      }
+      _tokenCache.remove(cacheKey);
+    }
+    return null;
+  }
+
+  String _getCacheKey(TokenRequestContext requestContext) =>
+      host + clientId + tenantId + requestContext.scopes.join(' ');
+
+  Future<http.Response> _postUrlFormBodyRequest(
+    Uri uri,
+    Map<String, String> formBody,
+  ) {
+    return _httpClient.post(
+      uri,
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: formBody,
+    );
+  }
+
+  Future<DeviceCodeInfo> _getDeviceCodeInfo(
+    TokenRequestContext requestContext,
+  ) async {
+    final substitutions = <String, dynamic>{};
+    substitutions['host'] = host;
+    substitutions['tenant_id'] = tenantId;
+    final uri = StdUriTemplate.expand(deviceCodeUriTemplate, substitutions);
+
+    final formBody = {
+      'client_id': clientId,
+      'scope': requestContext.scopes.join(' '),
+    };
+    final response = await _postUrlFormBodyRequest(Uri.parse(uri), formBody);
+
+    // TODO(baywet): use the refresh token if present and not expired.
+    if (response.statusCode != 200) {
+      throw Exception('Failed to get device code');
+    }
+    final responseBody = jsonDecode(response.body);
+    if (responseBody is! Map<String, dynamic>) {
+      throw Exception('Failed to parse device code response');
+    }
+    return DeviceCodeInfo.fromJson(responseBody);
+  }
+
+  Future<AccessToken> _pollForToken(DeviceCodeInfo codeInfo) async {
+    DeviceCodeTokenResponse? tokenResponse;
+    do {
+      var interval = codeInfo.interval ?? 0;
+      if (interval <= 0) {
+        interval = 3;
+      }
+      await Future<void>.delayed(Duration(seconds: interval));
+      tokenResponse = await _getTokenInformation(codeInfo);
```

Comment:
Design-wise, I'd suggest replacing this retry-loop with a custom `http.BasicClient` implementation and using it as a default value in the constructor of this class.

There, you could handle response errors and retries. Also it may be reused for other auth flows.

Reply:
why would you expect this aspect to be implemented at the transport layer when this is an applicative concern?

Reply:
not in this case, it's not a cross cutting concern but rather specific to this token acquisition flow https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-device-code

Reply:
I see. In any case, nowhere in the docs does it say anything about status code 400, did you find that through testing?

Reply:
> The device code flow is a polling protocol so errors served to the client must be expected prior to completion of user authentication.